src/dynamic_honey_badger.rs

```diff
@@ -0,0 +1,449 @@
+            for (id, tx_vec) in hb_batch.transactions {
+                let entry = batch.transactions.entry(id.clone());
+                let id_txs = entry.or_insert_with(Vec::new);
+                for tx in tx_vec {
+                    use self::Transaction::*;
+                    match tx {
+                        User(tx) => id_txs.push(tx),
+                        Change(change) => self.insert_vote(id.clone(), change),
+                        Propose(s_id, propose, sig) => self.handle_propose(&s_id, propose, &*sig)?,
+                        Accept(s_id, accept, sig) => self.handle_accept(&s_id, accept, &*sig)?,
+                    }
+                }
+            }
+            // If DKG completed, apply the change.
+            if let Some(change) = change.as_ref() {
+                if let Some((pub_key_set, sk)) = self.get_key_gen_output() {
+                    self.start_epoch = hb_batch.epoch + 1;
+                    self.apply_change(change, pub_key_set, sk)?;
+                    batch.change = Some(change.clone());
+                    changed = true;
+                }
+            }
+            // If a new change has a majority, restart DKG.
+            if self.current_majority() != change.as_ref() {
+                if let Some(change) = self.current_majority().cloned() {
+                    self.start_key_gen(change)?;
+                } else {
+                    self.key_gen = None;
+                }
+            }
+            self.output.push_back(batch);
+        }
+        self.messages
+            .extend_with_epoch(self.start_epoch, &mut self.honey_badger);
+        // If `start_epoch` changed, we can now handle some queued messages.
+        if changed {
+            let queue = mem::replace(&mut self.incoming_queue, Vec::new());
+            for (epoch, sender_id, msg) in queue {
+                self.handle_honey_badger_message(&sender_id, epoch, msg)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Restarts Honey Badger with a new set of nodes, and resets the Key Generation.
+    fn apply_change(
+        &mut self,
+        change: &Change<NodeUid>,
+        pub_key_set: PublicKeySet,
+        sk: ClearOnDrop<Box<SecretKey>>,
+    ) -> Result<()> {
+        self.votes.clear();
+        self.vote_counts.clear();
+        self.key_gen = None;
+        let mut all_uids = self.netinfo.all_uids().clone();
+        if !match *change {
+            Change::Remove(ref id) => all_uids.remove(id),
+            Change::Add(ref id, _) => all_uids.insert(id.clone()),
+        } {
+            debug!("No-op change: {:?}", change);
+        }
+        let netinfo = NetworkInfo::new(self.our_id().clone(), all_uids, sk, pub_key_set);
+        self.netinfo = netinfo.clone();
+        // TODO: Drop the buffer if this node was removed, to become an observer.
+        let buffer = self.honey_badger.drain_buffer();
+        self.honey_badger = HoneyBadger::new(Rc::new(netinfo), self.batch_size, buffer)?;
```

Comment:
Can you update the batch size here given that the number of nodes has increased?

Reply:
...or decreased.

Reply:
It depends on the exact requirements; there is still a [TODO](https://github.com/poanetwork/hbbft/blob/b3b3994ec1dd42d7a4765dc220d1e5470e35c6f5/src/honey_badger.rs#L46) in the code for that. I created issue https://github.com/poanetwork/hbbft/issues/88 for it.

For now, the code does the right thing in the situation of the issue's first bullet point, i.e. where we want a constant block size.